pkg/container/versioned/value.go

```diff
@@ -0,0 +1,337 @@
+	return v.version.CompareAndSwap(current, new)
+}
+
+type Closer func() bool
+
+// GetHandle returns a Handle for the current version and a closer for releasing old versions held
+// for this version.
+// 'name' is for debugging purposes only
+func (v *Manager) GetHandle(name string) (handle Handle, closer Closer) {
+	// version is monotonically increasing, so appending the current version to the end of the
+	// outstanding handles keeps it sorted
+	v.mutex.Lock()
+	version := v.version.Load()
+	handle = newHandle(version)
+	v.handles = append(v.handles, version)
+	v.mutex.Unlock()
+
+	return handle, func() bool {
+		if version == invalidVersion {
+			// 2nd call does nothing
+			return false
+		}
+
+		v.mutex.Lock()
+		n, found := slices.BinarySearch(v.handles, version)
+		if found {
+			v.handles = slices.Delete(v.handles, n, n+1)
+
+			// release memory for long empty slice
+			if len(v.handles) == 0 && cap(v.handles) > maxEmptyHandles {
+				v.handles = nil
+			}
+			version = invalidVersion
+		} else {
+			log.WithFields(logrus.Fields{
+				"version": handle.getVersion(),
+				"handle":  name,
+			}).Error("closer: Handle for version not found")
+
+			v.mutex.Unlock()
+			return false
+		}
+
+		// clean if needed
+		keepVersion := invalidVersion
+		if len(v.handles) > 0 {
+			keepVersion = v.handles[0]
+		}
+		// 'keepVersion' is the current version if there are no outstanding handles
+		if keepVersion == invalidVersion {
+			keepVersion = v.version.Load()
+		}
+		// call the cleaner for 'keepVersion' only if not already called for this
+		// 'keepVersion'
+		if keepVersion > v.cleanedVersion {
+			if v.cleaner != nil {
+				v.cleanedVersion = keepVersion
+				v.cleaner(keepVersion)
+			} else {
+				log.WithFields(logrus.Fields{
+					"version": handle.getVersion(),
+					"handle":  name,
+				}).Error("closer: cleaner function not set, stale data will accumulate.")
+			}
+		}
+
+		v.mutex.Unlock()
+		return true
+	}
+}
+
+// versionRange is a range from the first to last version.
+// 'past' can be atomically modified to a smaller value.
+type versionRange struct {
+	first Version       // first version this value is valid for
+	past  atomicVersion // first version this value is invalid for
+}
+
+func (r *versionRange) Contains(version Version) bool {
+	return r.first <= version && version < r.past.Load()
+}
+
+type valueNode[T any] struct {
+	versions versionRange
+	next     atomic.Pointer[valueNode[T]]
+	value    T
+}
+
+// Value is a container for versioned value, implemented as a lock-free linked list
+type Value[T any] struct {
+	// versionedNodes are non-overlapping and in sorted order
+	head atomic.Pointer[valueNode[T]]
+}
+
+// SetValueAtVersion adds the value with validity starting from 'version'.
+// All values are added with "infinite" validity, which is then truncated
+// when an overlapping entry is added, or a value is removed.
+//
+// 'version' must be later than any the current version visible to the readers.
```

Comment:
IMO panic would be the right thing to do, but we have a strict "no panics" policy. One way to resolve this would be to bring the given version forward and insert with a warning log?

Reply:
added an error return.